src/NegatableGuard.ts

```diff
@@ -0,0 +1,27 @@
+import {ReasonGuard, cloneGuard} from './ReasonGuard';
+
+export type NegatableGuard<FROM, TO extends FROM, N extends FROM = FROM> = ReasonGuard<FROM, TO> & {
+	negate: () => NegatableGuard<FROM, N, TO>
+}
+
+export type AlternativeGuard<LEFT, RIGHT> = NegatableGuard<LEFT|RIGHT, LEFT, RIGHT>;
+
+export const isNegatableGuard =
+	<FROM, TO extends FROM, N extends FROM = FROM>
+	(input: ReasonGuard<FROM, TO>): input is NegatableGuard<FROM, TO, N> =>
+		typeof input === 'function' && typeof (input as any).negate === 'function';
+
+export const buildNegatable =
+	<FROM, TO extends FROM, N extends FROM = FROM>
+	(input: () => ReasonGuard<FROM, TO>, negated: () => ReasonGuard<FROM, N>) => {
+		const negatableGuard: NegatableGuard<FROM, TO, N> = Object.assign(
+			cloneGuard(input()),
+			{negate: () => Object.assign(
```

Comment:
Formatting here is hard to read

Reply:
Alternative suggestions? This is an inherently wonky line of code and I've spent so much time staring at it that it all makes sense to me now, so I'm a poor judge.

Reply:
```typescript
export const buildNegatable =
	<FROM, TO extends FROM, N extends FROM = FROM>
	(input: () => ReasonGuard<FROM, TO>, negated: () => ReasonGuard<FROM, N>) => {
		const negatedGuard = Object.assign(cloneGuard(negated()), {
			negate() {
				return negatableGuard;
			},
		});
		const negatableGuard: NegatableGuard<FROM, TO, N> = Object.assign(cloneGuard(input()), {
			negate() {
				return negatedGuard;
			},
		});
		return negatableGuard;
	};
```
I didn't think the closure would be able to capture the later-declared const, but `vscode` isn't giving me any red squiggles.  If `vscode` is wrong, it's easy enough to make that a `let` declaration earlier.

The downside of this setup is that it eaglerly builds the negation now instead of on demand.  That can be undone at the expense of some indentation depth, though still more readable IMO than the original here:
```typescript
export const buildNegatable =
	<FROM, TO extends FROM, N extends FROM = FROM>
	(input: () => ReasonGuard<FROM, TO>, negated: () => ReasonGuard<FROM, N>) => {
		const negatableGuard: NegatableGuard<FROM, TO, N> = Object.assign(cloneGuard(input()), {
			negate() {
				return Object.assign(cloneGuard(negated()), {
					negate() {
						return negatableGuard;
					},
				});
			},
		});
		return negatableGuard;
	};
```

Reply:
I'm not a fan of the `negate() {` syntax vs. `negate: () =>` as it adds a layer of curly braces and a return. Happy to add more newlines and tabs.